### PR TITLE
New Plugin "RadSecProxy"

### PR DESCRIPTION
**Pull request by botboe**
Hi everyone, I hope I'm doing everything right... this is my first pull-request ;-)
I wrote a plugin for the radsecproxy-daemon, a secure radius proxy.

### CODE REVIEW

File: net/radsecproxy/+POST_INSTALL [hunk 1]
```diff
@@ -0,0 +1,4 @@
+if /usr/local/etc/rc.d/configd status > /dev/null; then
```
Comment (mimugmail) [inline net/radsecproxy/+POST_INSTALL:1]: Do you really need this file?

File: net/radsecproxy/Makefile [hunk 2]
```diff
@@ -0,0 +1,7 @@
+PLUGIN_NAME=		radsecproxy
+PLUGIN_VERSION=		1.0.2
```
Comment (mimugmail) [inline net/radsecproxy/Makefile:2]: You can start with 0.1 or 0.0.1 when merging to master

File: net/radsecproxy/Makefile [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+PLUGIN_COMMENT=		radsecproxy - a generic RADIUS proxy that provides both RADIUS UDP and TCP/TLS (RadSec) transport.
+PLUGIN_DEPENDS=		radsecproxy
+PLUGIN_MAINTAINER=	tobias@boehnert.dev
+
```
Comment (mimugmail) [inline net/radsecproxy/Makefile:6]: add PLUGIN_DEVEL= yes

File: net/radsecproxy/pkg-descr [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+a generic RADIUS proxy that in addition to usual RADIUS UDP
+transport, also supports TLS (RadSec), as well as RADIUS
+over TCP and DTLS. The aim is for the proxy to have
+sufficient features to be flexible, while at the same time
+to be small, efficient and easy to configure.
+
+Plugin Changelog
+================
+
+v1.0.2  Minor fixes
```
Comment (mimugmail) [inline net/radsecproxy/pkg-descr:13]: I'd add changelog when going to stable (1.0)

File: net/radsecproxy/src/opnsense/mvc/app/controllers/OPNsense/RadSecProxy/Api/ServiceController.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+use OPNsense\Base\ApiMutableServiceControllerBase;
+use OPNsense\Core\Backend;
+
+class ServiceController extends ApiMutableServiceControllerBase
+{
+    protected static $internalServiceClass = '\OPNsense\RadSecProxy\RadSecProxy';
+    protected static $internalServiceTemplate = 'OPNsense/RadSecProxy';
+    protected static $internalServiceEnabled = 'general.Enabled';
+    protected static $internalServiceName = 'radsecproxy';
+    protected function reconfigureForceRestart()
```
Comment (mimugmail) [inline net/radsecproxy/src/opnsense/mvc/app/controllers/OPNsense/RadSecProxy/Api/ServiceController.php:36]: Is this called somewhere? Maybe I missed it ...

File: net/radsecproxy/src/opnsense/mvc/app/controllers/OPNsense/RadSecProxy/forms/general.xml [hunk 6]
```diff
@@ -0,0 +1,51 @@
+<form>
+
+    <field>
+        <id>radsecproxy.general.Enabled</id>
```
Comment (mimugmail) [inline net/radsecproxy/src/opnsense/mvc/app/controllers/OPNsense/RadSecProxy/forms/general.xml:4]: Sometimes Enabled starts up- and lowcase, just cosmetic.

File: net/radsecproxy/src/opnsense/mvc/app/models/OPNsense/RadSecProxy/RadSecProxy.xml [hunk 7]
```diff
@@ -0,0 +1,372 @@
+<model>
+    <mount>//OPNsense/radsecproxy</mount>
+    <description>
+        RadSecProxy-Management
+    </description>
+    <version>1.0.2</version>
```
Comment (mimugmail) [inline net/radsecproxy/src/opnsense/mvc/app/models/OPNsense/RadSecProxy/RadSecProxy.xml:6]: Here you can also start with 0.X .. but technically no problem

File: net/radsecproxy/src/opnsense/mvc/app/models/OPNsense/RadSecProxy/RadSecProxy.xml [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+        </clients>
+
+        <servers>
+            <server type="ArrayField">
+
+                <identifier type="TextField">
+                    <Required>Y</Required>
+                    <mask>/^([0-9a-zA-Z_\-]){1,25}$/u</mask>
+                    <ValidationMessage>Should be a string between 1 and 25 characters whithout special characters.</ValidationMessage>
+
```
Comment (mimugmail) [inline net/radsecproxy/src/opnsense/mvc/app/models/OPNsense/RadSecProxy/RadSecProxy.xml:163]: Please remove the newline here

File: net/radsecproxy/src/opnsense/mvc/app/models/OPNsense/RadSecProxy/RadSecProxy.xml [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+    <mount>//OPNsense/radsecproxy</mount>
+    <description>
+        RadSecProxy-Management
+    </description>
+    <version>1.0.2</version>
+    <items>
+        <general>
+
+            <Enabled type="BooleanField">
+                <default>1</default>
```
Comment (mimugmail) [inline net/radsecproxy/src/opnsense/mvc/app/models/OPNsense/RadSecProxy/RadSecProxy.xml:11]: general.enabled should always start with 0 as default

File: net/radsecproxy/src/opnsense/mvc/app/models/OPNsense/RadSecProxy/RadSecProxy.xml [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+                    <Required>N</Required>
+                </description>
+                
+                <host type="TextField">
+                    <Required>Y</Required>
+                    <IpAllowed>Y</IpAllowed>
+                </host>
+
+                <port type="PortField">
+                </port>
```
Comment (mimugmail) [inline net/radsecproxy/src/opnsense/mvc/app/models/OPNsense/RadSecProxy/RadSecProxy.xml:182]: This looks a bit naked :) Required? Default?

File: net/radsecproxy/src/opnsense/mvc/app/models/OPNsense/RadSecProxy/RadSecProxy.xml [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+                    <Constraints>
+                        <check001>
+                            <type>UniqueConstraint</type>
+                            <ValidationMessage>Name already in use</ValidationMessage>
+                        </check001>
+                    </Constraints>
+                </name>
+
+                <description type="TextField">
+                </description>
```
Comment (mimugmail) [inline net/radsecproxy/src/opnsense/mvc/app/models/OPNsense/RadSecProxy/RadSecProxy.xml:271]: Same here. ..

File: net/radsecproxy/+POST_INSTALL (hunk 1 above)
Comment (botboe) [reply net/radsecproxy/+POST_INSTALL:1]: On my development-machine the menu-items won't show up without `configure_plugins`.
The `config restart` is part of the [tutorial](https://docs.opnsense.org/development/examples/helloworld.html#create-an-installable-plugin).

**Comment by botboe** [comment]
@mimugmail thanks for your feedback! How am I supposed to change my code respecting your comments and get them into this pull request? Or do I have to create a new pull request? Sorry, I'm kind of a git-noob ;-)

**Comment by mimugmail** [comment]
Just edit the files in your master branch, they'll get into here too :)

File: net/radsecproxy/src/opnsense/mvc/app/controllers/OPNsense/RadSecProxy/Api/ServiceController.php (hunk 5 above)
Comment (botboe) [reply net/radsecproxy/src/opnsense/mvc/app/controllers/OPNsense/RadSecProxy/Api/ServiceController.php:36]: Part of the [tutorial](https://docs.opnsense.org/development/examples/api_enable_services.html#update-the-service-controller): always stop daemon to reconfigure.

File: net/radsecproxy/src/opnsense/mvc/app/models/OPNsense/RadSecProxy/RadSecProxy.xml (hunk 10 above)
Comment (botboe) [reply net/radsecproxy/src/opnsense/mvc/app/models/OPNsense/RadSecProxy/RadSecProxy.xml:182]: Added "not required", defaults set by radsecproxy.

File: net/radsecproxy/src/opnsense/mvc/app/controllers/OPNsense/RadSecProxy/forms/general.xml (hunk 6 above)
Comment (botboe) [reply net/radsecproxy/src/opnsense/mvc/app/controllers/OPNsense/RadSecProxy/forms/general.xml:4]: Changed to lowercase.

**Comment by botboe** [comment]
> Just edit the files in your master branch, they'll get into here too :)
Fancy 👍  The changes are there...

**Comment by botboe** [comment]
Ok, everything should be fixed :-)

**Comment by botboe** [comment]
Hi @mimugmail, are you waiting for more input from me or do I have to do something like "re-request review"? Or didn't you just have time to check?

**Comment by mimugmail** [comment]
Didn't find the time yet .. week started with some trouble.

**Comment by botboe** [comment]
No problem, just wanted to be sure it's not my fault ;-)

**Comment by mimugmail** [comment]
I just installed it now:

- are custom fields really required? plugin wont be accepted with free-text
- I cant save any grid, does this work on your end?

**Comment by botboe** [comment]
I used custom fields for settings which won't be relevant in most cases so that the gui doesn't get overloaded. I took a lot of inspiration from the HAproxy-plugin which does the same (`customOptions type="TextField"`, [here](https://github.com/opnsense/plugins/blob/master/net/haproxy/src/opnsense/mvc/app/models/OPNsense/HAProxy/HAProxy.xml)).
The grid should be working normally, I already use the plugin in production. What does happen on saving? I just used the standard-methods of the mvc-framework.

**Comment by AdSchellevis** [comment]
We probably missed the option 4 years ago,  custom text inputs are security incidents waiting to happen for which a clear fix isn't available. In core we added some safeguards for legacy inputs that are still there https://github.com/opnsense/core/commit/d62015df1cdb0c0711b488bd66ced631b9a4f37b, the general rule of thumb is that we don't accept these types of inputs. (people can use config imports or template extensions to feed non standard options).

**Comment by botboe** [comment]
Ok, then I will pull the options I used the custom-block for into normal inputs. The support for some more "exotic" options will follow in a later release.

**Comment by mimugmail** [comment]
I code 200 as response but the popup doesn't close:

image

Also don't see any values. Is your version in production the same as here?

**Comment by mimugmail** [comment]
> Ok, then I will pull the options I used the custom-block for into normal inputs. The support for some more "exotic" options will follow in a later release.

You can hide them behind Advanced in forms so the UI doesn't get overloaded.

**Comment by botboe** [comment]
@mimugmail I wanted to remove the plugin completely to reproduce your problems. After removing it via the plugins-menu and reinstalling it again, the old database-entries like the clients are still there. How can I drop the plugin's database?

**Comment by mimugmail** [comment]
This is not relevant, just try to add more clients. 
If you want to clear it just delete all contencts before removing the plugin or remove the xml container from /conf/config.xml

**Comment by botboe** [comment]
> This is not relevant, just try to add more clients.

It is ;-) When clients are already configured, I can add a new without problems. If it is empty, it fails as you described. Interesting, I will investigate!

> If you want to clear it just delete all contencts before removing the plugin or remove the xml container from /conf/config.xml

Did the trick, thanks!

**Comment by mimugmail** [comment]
Oh, a new error I haven't seen yet :) 
Quite sure you'll find it

**Comment by botboe** [comment]
@mimugmail it should be fixed now, it was a wrong named default-value for two options which failed the validation. In my case these options were already set so the validation was ok.

**Comment by mimugmail** [comment]
Ok, got a running instance (running, not tested working), but you need to remove all custom field before this one gets merged.

**Comment by botboe** [comment]
Still reworking, please don't re-review now.

**Comment by botboe** [comment]
Everything done!
The rewrites-dialog contains freetext-fields, but they get splitted at the `\n` and get inserted each with a prefix into the config. I hope that is ok...

**Comment by botboe** [comment]
Hey @mimugmail, any update regarding this plugin?

**Comment by mimugmail** [comment]
@botboe I updated the branch und installed again, looks good so far. 
You could change the sorting of menues in alph. order. Quick huge plugin, I'll try to have a look at the code again but in general it seems to work.

**Comment by mimugmail** [comment]
@fichtner any way to merge this to master so wen can work on devel further?

**Review by fichtner — APPROVED** [review]

**Comment by fichtner** [comment]
Merged, thanks!

**Comment by botboe** [comment]
Hey Guys, thank you for merging. I didn't forget about the outstanding todos, but corona took its toll and I had some other priorities. I will create a minor-update in the future, but I think it'll be just cosmetics.

**Comment by botboe** [comment]
@fichtner, just for my understanding as a plugin-developer-newbie, when will the plugin be available through the opnsense-webgui for installation?

**Comment by fichtner** [comment]
> @fichtner any way to merge this to master so wen can work on devel further?

I was under the impression further work would be ongoing ;)

The plugin is available in the development version or for the release version via command line:

    # pkg install os-radsecproxy-devel

I don't know what should still be changed functionally, but at least from an organisational standpoint someone needs to let me know if this should be released as a stable plugin with a version 1.0 when the functional things have been decided and done. :)


Cheers,
Franco

**Comment by botboe** [comment]
> 
> 
> > @fichtner any way to merge this to master so wen can work on devel further?
> 
> I was under the impression further work would be ongoing ;)
> 
Ah ok, I misunderstood this point. There sure might be some improvements in the future, but more a kind of cosmetic changes when I have more time for development again.

> I don't know what should still be changed functionally, but at least from an organisational standpoint someone needs to let me know if this should be released as a stable plugin with a version 1.0 when the functional things have been decided and done. :)
> 
From the functional view everything seems to be good. I've got the plugin running successfully for several month and I think a 1.0 release is possible :-)

**Comment by fichtner** [comment]
@botboe ok then if @mimugmail agrees

**Comment by mimugmail** [comment]
@botboe would it be possible if you would write a piece of documentation to build a setup as the plugin is quite big.

**Comment by botboe** [comment]
@mimugmail at least some kind of high-level documentation might be possible. The detailed configuration itself is very individual per use-case like the one for [eduroam](https://www.dfn.de/fileadmin/1Dienstleistungen/Roaming/Einrichtung_von_radsecproxy.pdf).

**Comment by botboe** [comment]
@mimugmail could we do this in parallel? Release of version 1.0 and I will beginn writing some documentation? How can I contribute? PR for opnsense/docs?

**Comment by botboe** [comment]
@mimugmail could you give me some information regarding the questions above?

**Comment by mimugmail** [comment]
Yes you can clone the docs repo, copy a page which fits your needs and adjust with your plugin. E.g. wireguard howtos are quite detailed. It should help ppl understand what they can do with the plugin and how to achieve it (without prior knowledge).

**Comment by botboe** [comment]
Hi @mimugmail, I'm currently writing the [how-to](https://github.com/botboe/opnsense-docs-radsecproxy/blob/master/source/manual/how-tos/radsecproxy-eduroam.rst). Would you be so kind to release the plugin in parallel so that I can push it to my production-machine as soon as possible and to do a hardware-switch afterwards?

**Comment by botboe** [comment]
@fichtner can you give me an update? I got no repsonse from so far...

**Comment by fichtner** [comment]
@botboe what sort of update? 😊

**Comment by fichtner** [comment]
Maybe you are referring to this https://github.com/opnsense/changelog/blob/33f9d578089bac96b9829dadd2358b1e9b2d3f6d/community/21.7/21.7.r2#L39

**Comment by botboe** [comment]
> 
> 
> Maybe you are referring to this https://github.com/opnsense/changelog/blob/33f9d578089bac96b9829dadd2358b1e9b2d3f6d/community/21.7/21.7.r2#L39

Oh, I didn't see this... 😱 Awesome! 
I will tell you when the documentation is done and create a pull-request.

**Comment by fichtner** [comment]
sure, thanks 👍

**Pull request merged**